perfkitbenchmarker/scripts/object_storage_api_tests.py

```diff
@@ -518,6 +592,178 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
                   host_to_connect=host_to_connect)
 
 
+def RunMultiThreadedWorkersWriteResults(worker, worker_args,
+                                        storage_schema, host_to_connect=None):
+  """Run a worker function in many threads, then gather and print the results.
+
+  Args:
+    worker: either WriteWorker or ReadWorker. The worker function to call.
+    worker_args: a tuple. The arguments to pass to the worker function. The
+      result queue and stream number will be appended as the last two arguments.
+    storage_schema: The schema of the storage provider to use, e.g., "gs"
+    host_to_connect: An optional host endpoint to connect to.
+
+  The object size is given by FLAGS.object_size_kb. The number of
+  objects to transfer (per stream) is given by
+  FLAGS.objects_per_stream. The benchmark starts at FLAGS.start_time,
+  if given, or else immediately.
+
+  Raises:
+    LowAvailabilityError: when the storage provider has failed a high number of
+        our RW requests that exceeds a threshold (>5%), we raise this error
+        instead of collecting performance numbers from this run.
+  """
+
+  result_queue = Queue.Queue()
+
+  logging.info('Creating threads')
+  threads = [threading.Thread(target=worker,
+                              args=worker_args + (result_queue, i))
+             for i in xrange(FLAGS.num_streams)]
+  logging.info('Threads created. Starting threads.')
+  for thread in threads:
+    thread.start()
+  logging.info('Threads started. Joining threads.')
+  for thread in threads:
+    thread.join()
```

Comment:
The default is no timeout. I think that ideally each thread would send updates on its progress to the master thread, and we would kill a thread that wasn't making progress.

A less good but easier to implement solution is to poll threads to see if they are finished, and once some percentage are done, give the rest a fixed time interval to finish before we kill them. This would protect against the situation where most threads are okay but one or two are stuck.

An even less good alternative, which I think we should take for v1, is to let the user kill the benchmark if it takes too long.

What do you think of those ideas?